src/vsr/replica_format.zig

```diff
@@ -26,195 +22,287 @@ pub fn format(
-                    &self.write,
-                    wal_buffer[0..size],
-                    .wal_headers,
-                    wal_offset,
+            // Direct I/O requires the buffer to be sector-aligned. Unlike the Prepares above that
+            // require a buffer per prepare, since they are spread out with zeros inbetween, the
+            // headers zone is contiguous so a single buffer will do.
+            const headers_buffer = try arena.alignedAlloc(
+                u8,
+                constants.sector_size,
+                vsr.sector_ceil(constants.journal_size_headers),
+            );
+            @memset(headers_buffer, 0);
+
+            for (0..constants.journal_slot_count) |slot| {
+                const single_header_buffer =
+                    headers_buffer[slot * @sizeOf(Header.Prepare) ..][0..@sizeOf(Header.Prepare)];
+                const header, const prepare_offset = wal_header(
+                    cluster,
+                    slot,
+                    single_header_buffer,
                 );
-                self.formatting = true;
-                while (self.formatting) storage.run();
-                wal_offset += size;
+                _ = prepare_offset;
+
+                assert(!stdx.zeroed(single_header_buffer));
+                assert(stdx.zeroed(headers_buffer[(slot + 1) * @sizeOf(Header.Prepare) ..]));
+                assert(header.valid_checksum());
+
+                if (header.op == 0) {
+                    assert(header.operation == .root);
+                } else {
+                    assert(header.operation == .reserved);
+                }
             }
-            // There are no headers left to write.
-            assert(format_wal_headers(cluster, wal_offset, wal_buffer) == 0);
+
+            storage.write_sectors(
+                write_sectors_callback,
+                &self.writes[self.writes_pending],
+                headers_buffer,
+                .wal_headers,
+                0,
+            );
+            self.writes_pending += 1;
         }
 
-        fn format_replies(
+        fn queue_format_replies(
             self: *ReplicaFormat,
-            allocator: std.mem.Allocator,
+            arena: std.mem.Allocator,
             storage: *Storage,
         ) !void {
-            assert(!self.formatting);
+            assert(!self.formatting and !self.formatting_superblock);
 
             // Direct I/O requires the buffer to be sector-aligned.
             const message_buffer =
-                try allocator.alignedAlloc(u8, constants.sector_size, constants.message_size_max);
-            defer allocator.free(message_buffer);
+                try arena.alignedAlloc(
+                    u8,
+                    constants.sector_size,
+                    constants.message_size_max * constants.clients_max,
+                );
+
             @memset(message_buffer, 0);
 
             for (0..constants.clients_max) |slot| {
```

Comment:
Why do we need a loop here? Isn't this effectively just a single big write? Do we need to write anything here at all, given that we no longer try to zero-out the entire file (for wal prepares, we only writing headers).

Reply:
> Do we need to write anything here at all

No! And we don't need to write anything for the padding either!